README.md
# nuflakbrr-speedtest.vercel.app

Create React Website Speed Test App.

## Authors

Contributors names and contact info

Original scripts by <br />
Rahmat Agung Julians | [@rahmatagungj](https://github.com/rahmatagungj)

Convert scripts from TypeScript to JavaScript by <br />
Naufal Akbar Nugroho | [@kbrnugroho](https://instagram.com/kbrnugroho)

## Version History

- 1.0.0
  - [Initial Release](CHANGELOG.md)


CHANGELOG.md
# Change log

## v1.0.0 — Aug. 15, 2021

Initial release.

**Added:**

- Readme
- Change log
- License


src/pages/Details.js
import React from "react";
import { siteContext } from "../contexts/SiteContexts";
import { Redirect } from "react-router-dom";
import AuditsCard from "../components/AuditsCard";
import MainLayout from "../layouts/MainLayout";

const Details = () => {
  const { siteData } = React.useContext(siteContext);

  const auditsToShow = [
    "speed-index",
    "largest-contentful-paint",
    "interactive",
    "first-contentful-paint",
    "first-meaningful-paint",
    "largest-contentful-paint-element",
    "uses-long-cache-ttl",
    "cumulative-layout-shift",
    "layout-shift-elements",
  ];

  return (
    <MainLayout>
      {!siteData && <Redirect to="/" />}
      <div className="px-6 py-6 bg-blue-600 dark:bg-gray-800 md:py-12 md:px-12 lg:py-16 lg:px-16 xl:flex xl:items-center">
        <div className="xl:w-0 xl:flex-1">
          <h2 className="break-words text-xl md:text-2xl leading-8 font-extrabold tracking-tight text-white sm:text-3xl sm:leading-9">{siteData?.lighthouseResult.finalUrl}</h2>
          <p className="mt-3 max-w-3xl text-md md:text-lg leading-6 text-indigo-200">{siteData?.lighthouseResult.userAgent}</p>
          <p className="text-xs md:text-sm leading-5 text-indigo-200">Fetch Data at {siteData?.lighthouseResult.fetchTime}</p>
          <p className="text-xs md:text-sm leading-5 text-indigo-200">Lighthouse Version {siteData?.lighthouseResult.lighthouseVersion}</p>
        </div>
        <div className="mt-8 sm:w-full sm:max-w-md xl:mt-0 xl:ml-8">
          <img src={siteData?.lighthouseResult.audits["final-screenshot"].details.data} width={400} alt="final screenshot" />
        </div>
      </div>
      <div className="flex justify-center mx-auto">
        <div className="p-0 md:p-5 m-5 flex flex-col md:flex-row flex-wrap justify-start">
          {siteData &&
            auditsToShow.map((name, idx) => (
              <AuditsCard key={idx + name} times={siteData?.lighthouseResult.audits[name].displayValue} title={siteData?.lighthouseResult.audits[name].title} description={siteData?.lighthouseResult.audits[name].description} />
            ))}
        </div>
      </div>
    </MainLayout>
  );
};

export default Details;


src/pages/Home.js
import React from "react";
import MainLayout from "../layouts/MainLayout";
import axios from "axios";
import { siteContext } from "../contexts/SiteContexts";
import { useHistory } from "react-router-dom";

const API_KEY = process.env.REACT_APP_API_KEY;

const Home = () => {
  const { setSiteData } = React.useContext(siteContext);
  const [isLoading, setIsLoading] = React.useState(false);
  const [isError, setIsError] = React.useState(false);
  const url = React.useRef();
  const history = useHistory();
  const HandleCheck = async () => {
    setIsLoading(true);
    if (isError) setIsError(false);
    try {
      const response = await axios.get(`https://www.googleapis.com/pagespeedonline/v5/runPagespeed?url=${url.current?.value}&key=${API_KEY}`);
      setSiteData(response.data);
      setIsLoading(false);
      history.push(`/${url.current?.value}`);
    } catch (e) {
      setIsError(true);
      setIsLoading(false);
    }
  };

  return (
    <MainLayout>
      <div className="flex h-screen bg-gray-900">
        <section className="m-auto">
          <div className="container flex flex-col px-5 py-8 mx-auto lg:items-center">
            <div className="flex flex-col w-full mb-8 text-left lg:text-center justify-center">
              <h1 className="mx-auto mb-6 text-2xl font-semibold leading-none tracking-tighter text-gray-100 lg:w-5/6 sm:text-5xl title-font">Tes Kecepatan Website</h1>
              <p className="mx-auto text-base font-normal leading-relaxed text-gray-200 lg:w-5/6 text-center">Semakin cepat pemuatan website, semakin tinggi nilai website anda di penilaian Google.</p>
              <div className="mt-10" />
              {isError && (
                <div className="p-2 justify-center flex flex-row">
                  <div className="inline-flex items-center leading-none text-white rounded-lg p-2 shadow text-teal text-sm">
                    <span className="inline-flex bg-red-600 text-white rounded-full h-6 px-3 justify-center items-center">Gagal</span>
                    <span className="inline-flex px-2">Situs tidak ditemukan</span>
                  </div>
                </div>
              )}
              <div className="flex flex-col md:flex-row md:items-start w-full mx-auto justify-center lg:w-5/6">
                <div className="relative md:mr-4 text-left lg:w-full xl:w-1/2 md:w-full">
                  <input
                    type="text"
                    id="urlsite"
                    placeholder="Masukan URL Halaman Situs"
                    name="urlsite"
                    ref={url}
                    disabled={isLoading}
                    className="flex-grow w-full px-4 py-2 mb-4 md:mr-4 text-base text-black transition duration-650 ease-in-out transform rounded-lg bg-blueGray-200 focus:outline-none focus:border-purple-500 sm:mb-0 focus:bg-white focus:shadow-outline focus:ring-2 ring-offset-current ring-offset-2"
                  />
                </div>
                <button
                  onClick={HandleCheck}
                  disabled={isLoading}
                  className={`${
                    isLoading && "cursor-not-allowed"
                  } mx-auto md:mx-0 w-32 text-center justify-center md:w-20 flex items-center px-6 py-2 mt-auto font-semibold text-white transition duration-500 ease-in-out transform bg-blue-600 rounded-lg hover:bg-blue-700 focus:shadow-outline focus:outline-none focus:ring-2 ring-offset-current ring-offset-2`}
                >
                  {isLoading ? "Memuat" : "Cek"}
                </button>
              </div>
            </div>
          </div>
        </section>
      </div>
    </MainLayout>
  );
};

export default Home;
